_source/_posts/2022-07-21-angular-security-xss.md

```diff
@@ -0,0 +1,402 @@
+
+As you can see, there are many vectors for the vulnerability. Many of these sinks have legitimate use cases when building dynamic web applications. Since these sinks are necessary for web app functionality, we must use trusted data by escaping and sanitizing it.
+
+> This is still a high-level summary. The fine print is that XSS vulnerabilities can be pretty sneaky with the different ways the attack can occur.
+
+There are different XSS attacks, each with a slightly different attack vector. We'll briefly cover how three attacks work.
+
+### Stored XSS
+
+In this flavor of XSS, the attack is persisted somewhere like in a database. Stored XSS is the example recapped above where an agitator's terrible comment with the `script` tag persists in the database and then ruins someone else's day by showing the unfriendly comment in an alert.
+
+### Reflected XSS
+
+This attack is where the malicious code sneaks in through the HTTP request, usually through URL parameters. Suppose the K-Drama site takes a search term via a URL parameter, such as
+
+```
+https://myfavekdramas.com/dramas?search=crash+landing+on+you
+```
+
+The site then takes the search terms and displays them back to the user while calling out to the back-end to run the search.
+
+But what if an agitator constructs a URL like this?
+
+```
+https://myfavekdramas.com/dramas?search=<img src=1 onerror="alert('Doh!')"/>
+```
+
+You may think that you'd never navigate to a link like that! Who would?! But let's remember, in a previous post, you _did_ click the link in your spam email to send money to your high school sweetheart. Pointing this out is not judgment but merely pointing out the past behavior of clicking fishy links. Also, agitators are pretty tricksy. They might use URL shorteners to obscure it.
+
+### DOM-based XSS
+
+In this attack, the agitator takes advantage of Web APIs and occurs entirely within the SPA. This attack is pretty much identical to Reflective XSS. 
+
+Let's say our application depends on an external resource—the app embeds an `<iframe>` for showing trailers for the K-Dramas and sets the `iframe`'s `src` attribute to an external site. So our code might look like this.
+
+```html
+<iframe src="{resourceURL}" />
+```
+
+We usually call out to the third-party service to get the URLs for the source, but agitators have infiltrated this third-party service and now control the resource URLs returned, making our code look like this.
+
+```html
+<iframe src="javascript:alert('Boo!')" />
+```
+
+Well, darn it, we've got some problems.
+
+## XSS Support in Angular
+
+Fortunately, Angular has a lot of built-in security protections. Angular treats all values as suspicious and untrusted by default, which is incredibly helpful because the framework automatically guards us against unintentionally creating vulnerabilities in our applications. Angular automatically removes any `script` tags so we won't have to worry about the original hypothentical example.
```

Comment:
```suggestion
Fortunately, Angular has a lot of built-in security protections. It treats all values as suspicious and untrusted by default, which is incredibly helpful because the framework automatically guards us against unintentionally creating vulnerabilities in our applications. Angular automatically removes any `script` tags so we won't have to worry about the original hypothentical example.
```
avoids the repetition of "Angular" to open consecutive sentences